gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/reactive/handlers/api/security/handler/SecurityHandler.java

```diff
@@ -0,0 +1,173 @@
+import io.gravitee.gateway.reactive.api.policy.SecurityPolicy;
+import io.gravitee.repository.management.api.SubscriptionRepository;
+import io.gravitee.repository.management.api.search.SubscriptionCriteria;
+import io.gravitee.repository.management.model.Subscription;
+import io.reactivex.Completable;
+import io.reactivex.Single;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Optional;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link SecurityHandler} allows to wrap a {@link io.gravitee.gateway.reactive.api.policy.Policy} implementing {@link SecurityPolicy} and make it working in a security chain.
+ * Security handler is responsible to
+ * <ul>
+ *     <li>Check if a policy can handle the security or not</li>
+ *     <li>Check the eventual selection rule matches (useful when dealing with multiple plans relying on the same security scheme such as <code>Authorization: bearer xxx</code> for JWT and OAuth)</li>
+ *     <li>Check if the policy requires to validate there is an associated subscription or not and validate the subscription accordingly</li>
+ *     <li>Invoke the <code>onSubscriptionInvalid</code> method when necessary</li>
+ * </ul>
+ *
+ * @author Jeoffrey HAEYAERT (jeoffrey.haeyaert at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class SecurityHandler {
+
+    private static final Logger log = LoggerFactory.getLogger(SecurityHandler.class);
+
+    protected static final String CONTEXT_ATTRIBUTE_CLIENT_ID = "oauth.client_id";
+    protected static final Single<Boolean> TRUE = Single.just(true), FALSE = Single.just(false);
+
+    private final Plan plan;
+    private final SecurityPolicy policy;
+    private final String selectionRule;
+
+    SecurityHandler(@Nonnull final Plan plan, @Nonnull final SecurityPolicy policy) {
+        this.plan = plan;
+        this.policy = policy;
+        this.selectionRule = getSelectionRule(plan.getSelectionRule());
+    }
+
+    private String getSelectionRule(String selectionRule) {
+        if (selectionRule == null) {
+            return null;
+        }
+
+        if (selectionRule.startsWith("#")) {
+            // Backward compatibility. In V3 mode selection rule EL expression based can be defined with "#something" while it is usually defined with "{#something}" everywhere else.
+            return "{" + selectionRule + "}";
+        }
+        return selectionRule;
+    }
+
+    /**
+     * Make sure the policy can handle the current request.
+     * It basically invokes the <code>canHandle</code> method on the policy, then check if the optional selection rule matches.
+     *
+     * @param ctx the current execution context.
+     * @return <code>true</code> if this security handler can handle the request, <code>false</code> otherwise.
+     */
+    public Single<Boolean> canHandle(RequestExecutionContext ctx) {
```

Comment:
```suggestion
    public Single<Boolean> match(RequestExecutionContext ctx) {
```

Reply:
@jhaeyaert one missing change there

Reply:
Forgot this one indeed....

`SecurityHandler` -> `canHandle()` seems logical to me... not a bit fan to rename it `match`

Reply:
After a quick chat we decided to rename it `SecurityPlan` with `canExecute` and execute `methods`